src/traits/block_storage.rs

```diff
@@ -0,0 +1,37 @@
+//! Abstract storage type for storing DA proposals and VID shares
+//!
+//! This modules provides the [`BlockStorage`] trait.
+//!
+
+use async_trait::async_trait;
+use snafu::Snafu;
+
+use crate::{
+    data::{DAProposal, VidDisperse},
+    message::Proposal,
+};
+
+use super::node_implementation::NodeType;
+
+#[derive(Snafu, Debug)]
+#[snafu(visibility(pub))]
+pub enum BlockStorageError {
+    #[snafu(display("Failed to store the proposal due to an internal error"))]
+    StoreError,
+
+    #[snafu(display("Failed to retrieve the proposal due to an internal error"))]
+    RetrieveError,
+}
+
+#[derive(Clone, Debug)]
+pub enum ProposalType<TYPES: NodeType> {
+    DAProposal(Proposal<TYPES, DAProposal<TYPES>>),
+    VidDisperse(Proposal<TYPES, VidDisperse<TYPES>>),
+}
+
+/// Abstraction for storing the contents of DA proposals and VID shares.
+#[async_trait]
+pub trait BlockStorage<TYPES: NodeType>: Send + Sync + Clone {
+    async fn append(&self, proposal: &ProposalType<TYPES>) -> Result<(), BlockStorageError>;
```

Comment:
What's the rationale for taking an enum here? Feels like it would be cleaner to have separate methods `append_da` and `append_vid`, since they're pretty much independent stores?

Reply:
Yeah that seems reasonable to me.